store/tikv/txn_committer.go

```diff
@@ -84,51 +84,70 @@ func (c *txnCommitter) primary() []byte {
 	return c.keys[0]
 }
 
-func (c *txnCommitter) iterKeysByRegion(keys [][]byte, f func(RegionVerID, [][]byte) error) error {
-	groups := make(map[RegionVerID][][]byte)
-	var primaryRegionID RegionVerID
-	var lastRegion *Region
-	for _, k := range keys {
-		var region *Region
-		if lastRegion != nil && lastRegion.Contains(k) {
-			region = lastRegion
-		} else {
-			var err error
-			region, err = c.store.regionCache.GetRegion(k)
-			if err != nil {
-				return errors.Trace(err)
-			}
-			lastRegion = region
-		}
-		id := region.VerID()
-		if bytes.Compare(k, c.primary()) == 0 {
-			primaryRegionID = id
-		}
-		groups[id] = append(groups[id], k)
+// iterKeys groups keys into batches, then applies `f` to the them. If the flag
+// asyncNonPrimary is set, it will return as soon as the primary batch is
+// processed.
+func (c *txnCommitter) iterKeys(keys [][]byte, f func(batchKeys) error, sizeFn func([]byte) int, asyncNonPrimary bool) error {
+	if len(keys) == 0 {
+		return nil
 	}
+	groups, firstRegion, err := c.store.regionCache.GroupKeysByRegion(keys)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	firstIsPrimary := bytes.Equal(keys[0], c.primary())
 
+	var batches []batchKeys
 	// Make sure the group that contains primary key goes first.
-	if primaryRegionID.id != 0 {
-		if err := f(primaryRegionID, groups[primaryRegionID]); err != nil {
+	if firstIsPrimary {
```

Comment:
It happens when we got a RegionError while sending a batch (which may not contain the primary key), then we have to call `iterKeys` to re-split the batch.